### PR TITLE
arch/arm64: Fix ARM64 build

### LINKED ISSUE #349: Kata-env and Kata-check are not architecture-independent
# Description of problem
For now, the functionality of kata-check and kata-env are missing in Arm architecture. Error outputs are as follows:

```
# github.com/kata-containers/runtime/cli
./kata-check.go:250:65: undefined: archKernelParamHandler
./kata-check.go:271:27: undefined: archRequiredCPUFlags
./kata-check.go:272:27: undefined: archRequiredCPUAttribs
./kata-check.go:273:27: undefined: archRequiredKernelModules
./kata-check.go:285:10: undefined: archHostCanCreateVMContainer
./kata-env.go:179:26: undefined: archRequiredCPUFlags
./kata-env.go:180:26: undefined: archRequiredCPUAttribs
./kata-env.go:181:26: undefined: archRequiredKernelModules
Makefile:310: recipe for target '/home/Wei/penny/go/src/github.com/kata-containers/runtime/kata-runtime' failed
make: *** [/home/Wei/penny/go/src/github.com/kata-containers/runtime/kata-runtime] Error 2

```
It could come from following aspects:
1. Struct `vmContainerCapableDetails` was architecture-relevant, for example, there were no tag showing virtualization capability in /proc/cpuinfo in aarch64, so variable `requiredCPUFlags` could be no use. `flag` filed in /proc/cpuinfo may only amd64-specific.
3. Struct `CPUInfo` was also architecture-relevant, e.g. variable vendor and model are meaningless for arm architecture. Because func `getCPUDetails()` is related to `CPUInfo`, it may not be included in universal file cli/utils.go 
4. Func `hostIsVMContainerCapable()` is only suitable for amd64 to check host's capability of creating a VM container

We think that we could move the architecture-relevant variable and func to files with specific architecture suffixes. e.g. func `getCPUFlags`, `checkCPUFlags`, `hostIsVMContainerCapable` and const `cpuFlagsTag` should be transferred to file `kata-check_amd64.go`. 

@bergwolf @gnawux @jodh-intel @weichen81

### PR DESCRIPTION
**Pull request by jodh-intel**
Fix ARM64 build which silently broken (as we still don't have an ARM CI).

Fixes #349.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

### CODE REVIEW

**Comment by jodh-intel** [comment]
Related: https://github.com/kata-containers/ci/issues/30.

**Comment by katabuilder** [comment]
PSS Measurement:
Qemu: 161955 KB
Proxy: 6686 KB
Shim: 10978 KB

Memory inside container:
Total Memory: 2045972 KB
Free Memory:  1997864 KB

**Comment by devimc** [comment]
lgtm

**Comment by grahamwhaley** [comment]
Strange CRI-O fail on 16.04 CI:
```
ok 22 ctr caps drop
ok 23 run ctr with image with Config.Volumes
ok 24 # skip (This is not working (Issue https://github.com/kata-containers/agent/issues/138)) ctr oom
ok 25 ctr /etc/resolv.conf rw/ro mode
ok 26 ctr create with non-existent command
ok 27 ctr create with non-existent command [tty]
ok 28 # skip (This is not working (Issue https://github.com/kata-containers/agent/issues/138)) ctr update resources
ok 29 ctr correctly setup working directory
ok 30 ctr execsync conflicting with conmon env
ok 31 # skip (This is not working (Issue https://github.com/kata-containers/agent/issues/138)) ctr resources
Makefile:35: recipe for target 'crio' failed
make: *** [crio] Error 1
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build was aborted  : False
Match found for :Build step 'Execute shell' marked build as failure : True
Logical operation result is TRUE
Running script  : #!/bin/bash
```
/cc @chavafg

**Comment by chavafg** [comment]
Yep, this week we have seen some cri-o random failures. We are documenting the failures on: https://github.com/kata-containers/tests/issues/361
Restarted the job.

File: cli/kata-check_arm64.go [hunk 1]
```diff
@@ -5,6 +5,37 @@
 
 package main
 
+import "github.com/sirupsen/logrus"
+
+const (
+	cpuFlagsTag        = "Features"
+	archCPUVendorField = "CPU implementer"
```
Comment (Weichen81) [inline cli/kata-check_arm64.go]: About the implementer here, I think Arm and x86 may have some differences here. The CPU implementer here for Arm means the CPU core IP implementer, not the SoC vendor. For example, if Apple and Qualcomm use the Arm's IP implementation to design their SoCs (In actually, they use their own implementations :) ), the CPU implementer here would be the same for these two SoC vendors. So, if the archCPUVendorField presents the CPU core vendor, I think you can use CPU implementer here. How does x86 use the archCPUVendorField?

File: cli/kata-check_ppc64le.go [hunk 2]
```diff
@@ -37,11 +44,6 @@ func archHostCanCreateVMContainer() error {
 // hostIsVMContainerCapable checks to see if the host is theoretically capable
 // of creating a VM container.
 func hostIsVMContainerCapable(details vmContainerCapableDetails) error {
-	_, err := getCPUInfo(details.cpuInfoFile)
-	if err != nil {
-		return err
-	}
-
 	count, err := checkKernelModules(details.requiredKernelModules, archKernelParamHandler)
```
Comment (Weichen81) [inline cli/kata-check_ppc64le.go]: I have a concern about modules check. What will happen here, if we are running with a built-in kernel : ) ?

File: cli/kata-check_arm64.go (hunk 1 above)
Comment (jodh-intel) [reply cli/kata-check_arm64.go]: On `x86_64`, this information is used to display some useful info in the output of `kata-env`:

```
  [Host.CPU]
    Vendor = "GenuineIntel"
    Model = "Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz"
```

I wasn't really sure how to make that relevant for ARM, so any advice would be much appreciated ;)

File: cli/kata-check_ppc64le.go (hunk 2 above)
Comment (jodh-intel) [reply cli/kata-check_ppc64le.go]: That's handled :)

**Comment by Weichen81** [comment]
@jodh-intel commented on this pull request.

________________________________

In cli/kata-check_arm64.go<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fkata-containers%2Fruntime%2Fpull%2F362%23discussion_r192436290&data=02%7C01%7C%7C20e6202a55c54413dbec08d5c7d63383%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636634645150116633&sdata=5ur8nGTFS6BUNlchjqYFs0sq%2FyEEPxuOlr24byboutU%3D&reserved=0>:

> @@ -5,6 +5,37 @@

 package main

+import "github.com/sirupsen/logrus<http://github.com/sirupsen/logrus>"
+
+const (
+       cpuFlagsTag        = "Features"
+       archCPUVendorField = "CPU implementer"


On x86_64, this information is used to display some useful info in the output of kata-env:

  [Host.CPU]
    Vendor = "GenuineIntel"
    Model = "Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz"


I wasn't really sure how to make that relevant for ARM, so any advice would be much appreciated ;)

Ah, in this case, I think we can use “CPU implementer” here. But this still needs further formatting. But don’t worry, We can handle it after this PR merged ; )

**Comment by jodh-intel** [comment]
Great - thanks @Weichen81!

**Comment by jodh-intel** [comment]
Hi @Weichen81, @Pennyzct - are you happy for this PR to be merged?

**Comment by Pennyzct** [comment]
Hi~ @jodh-intel .LGTM😊. thanks for all! It run successfully in Arm server. I have already pulled request for refining CPUInfo😝. #368

**Comment by Weichen81** [comment]
Of course, @jodh-intel , thank you very much!

**Comment by jodh-intel** [comment]
Thanks @Pennyzct, @Weichen81 - sorry, my browser cached the page so I see now, it had already been merged :)

**Pull request merged**